crates/store/src/db/models/queries/accounts.rs

```diff
@@ -149,6 +149,68 @@ pub(crate) fn select_accounts_by_id(
     Ok(account_infos)
 }
 
+pub(crate) fn select_account_vault_assets(
+    conn: &mut SqliteConnection,
+    account_id: AccountId,
+    block_from: BlockNumber,
+    block_to: BlockNumber,
+) -> Result<(BlockNumber, Vec<AccountVaultValue>), DatabaseError> {
+    use schema::account_vault_assets as t;
+
+    // SELECT block_num, faucet_id, delta
+    // FROM account_vault_assets
+    // WHERE account_id = ?
+    //   AND block_num >= ?
+    //   AND block_num <= ?
+    // ORDER BY block_num ASC
+    // LIMIT ROW_LIMIT;
+
+    // TODO: These limits should be given by the protocol.
+    // See miden-base/issues/1770 for more details
+    const MAX_PAYLOAD_BYTES: usize = 2 * 1024 * 1024; // 2 MB
+    const ROW_OVERHEAD_BYTES: usize = core::mem::size_of::<Word>() * 2; // vault key word + asset word
+    const ROW_LIMIT: usize = (MAX_PAYLOAD_BYTES / ROW_OVERHEAD_BYTES) + 1;
+
+    if !account_id.is_public() {
+        return Err(DatabaseError::AccountNotPublic(account_id));
+    }
+
+    if block_from > block_to {
+        return Err(DatabaseError::InvalidBlockRange { from: block_from, to: block_to });
+    }
+
+    let mut raw: Vec<(i64, Vec<u8>, Option<Vec<u8>>)> =
+        SelectDsl::select(t::table, (t::block_num, t::vault_key, t::asset))
+            .filter(
+                t::account_id
+                    .eq(account_id.to_bytes())
+                    .and(t::block_num.ge(block_from.to_raw_sql()))
+                    .and(t::block_num.le(block_to.to_raw_sql())),
+            )
+            .order(t::block_num.asc())
+            .limit(i64::try_from(ROW_LIMIT).expect("should fit within i64"))
+            .load::<(i64, Vec<u8>, Option<Vec<u8>>)>(conn)?;
+
+    // Discard the last block in the response (assumes more than one block may be present)
+    let mut last_block_included = block_to.checked_sub(1).unwrap_or_default();
```

Comment:
I don't understand why we're discarding the last block?

Reply:
The idea is that if we get enough rows where we know there is at minimum 2 blocks involved, then we can always discard the last one (because it would likely not contain all updates for that block) and keep all the other updates, which would contain, for all blocks there, all updates and not partial data. This is the reason for calculating the overhead based on the limit of updates in a block (which I made up, for now, but they should be defined by the protocol) and adding 1 on top of that.